keys.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// CheckKeys checks the keys returned from a server to make sure they are valid.
+// If the checks pass then also return a map of key_id to Ed25519 public key and a list of SHA256 TLS fingerprints.
+func CheckKeys(serverName string, timeNowMs uint64, keys ServerKeys, connState *tls.ConnectionState) (
+	checks KeyChecks, ed25519Keys map[string]Base64String, sha256Fingerprints []Base64String,
+) {
+	checks.MatchingServerName = serverName == keys.ServerName
+
+	ed25519Keys = checkVerifyKeys(keys, &checks)
+	sha256Fingerprints = checkTLSFingerprints(keys, &checks)
+
+	checks.FutureValidUntilTS = timeNowMs < keys.ValidUntilTS
+
+	checks.AllChecksOK = (checks.MatchingServerName &&
+		checks.FutureValidUntilTS &&
+		checks.HasEd25519VerifyKey &&
+		*checks.ValidEd25519VerifyKey &&
+		*checks.MatchingEd25519Signature &&
+		checks.HasTLSFingerprint &&
+		*checks.ValidSHA256TLSFingerprint)
+
+	// Only check the fingerprint if we have the TLS connection state.
+	if connState != nil {
+		// Check the peer certificates.
+		matches := checkFingerprint(connState, sha256Fingerprints)
+		checks.MatchingTLSFingerprint = &matches
+		checks.AllChecksOK = checks.AllChecksOK && matches
+	}
+
+	if !checks.AllChecksOK {
+		sha256Fingerprints = nil
+		ed25519Keys = nil
+	}
+	return
+}
+
+func checkFingerprint(connState *tls.ConnectionState, sha256Fingerprints []Base64String) bool {
+	if len(connState.PeerCertificates) == 0 {
+		return false
+	}
+	cert := connState.PeerCertificates[0]
+	digest := sha256.Sum256(cert.Raw)
+	for _, fingerprint := range sha256Fingerprints {
+		if bytes.Compare(digest[:], fingerprint) == 0 {
+			return true
+		}
+	}
+	return false
+}
+
+func checkVerifyKeys(keys ServerKeys, checks *KeyChecks) map[string]Base64String {
+	validEd25519VerifyKey := true
+	matchingEd25519Signature := true
+	verifyKeys := map[string]Base64String{}
+	for keyID, keyData := range keys.VerifyKeys {
+		algorithm := strings.SplitN(keyID, ":", 2)[0]
+		publicKey := keyData.Key
+		if algorithm == "ed25519" {
+			if len(publicKey) != 32 {
+				validEd25519VerifyKey = false
+				continue
+			}
+			if err := VerifyJSON(keys.ServerName, keyID, []byte(publicKey), keys.Raw); err != nil {
+				matchingEd25519Signature = false
+				continue
+			}
+			checks.HasEd25519VerifyKey = true
+			checks.ValidEd25519VerifyKey = &validEd25519VerifyKey
+			checks.MatchingEd25519Signature = &matchingEd25519Signature
+			verifyKeys[keyID] = publicKey
+		}
+	}
+	return verifyKeys
+}
+
+func checkTLSFingerprints(keys ServerKeys, checks *KeyChecks) []Base64String {
+	var fingerprints []Base64String
+	validSHA256TLSFingerprint := true
+	for _, fingerprint := range keys.TLSFingerprints {
+		if len(fingerprint.SHA256) != sha256.Size {
+			validSHA256TLSFingerprint = false
+			continue
+		}
+		checks.HasTLSFingerprint = true
+		checks.ValidSHA256TLSFingerprint = &validSHA256TLSFingerprint
+		fingerprints = append(fingerprints, fingerprint.SHA256)
+	}
```

Comment:
Aha, yes that makes sense. This still fails if `keys.TLSFingerprints` has 1 entry: [bad] though.